### PR TITLE
DISCUSSION DRAFT: Facility to support dependency-free JSON, System.Text.Json, and Newtonsoft.Json

### PR DESCRIPTION
**Pull request by sblom**
Code to illustrate my thoughts on #65

### Current Regressions
- Lost support for .NET Standard 1.x (even 1.3) because of reflection requirements in `TinyJson`. This can be overcome.
- Many tests currently failing. I had almost all of the conformance cases passing with `OmniJson` in a previous commit. I'll go back and make a pass to fix up broken cases once I get the Newtonsoft implementation finished up.

### Overview
This change removes the `Newtonsoft.Json` dependency from the main project, and replaces it with a home-built facility called 'OmniJson' that provides a similar interface which supports dependency-free JSON by way of `TinyJson`, or can be enhanced with additional NuGet mix-ins (possibly called `json-ld.net.Newtonsoft` and `json-ld.net.SystemTextJson`) that would override `OmniJsonToken` et al, and provide API methods that expect and return Newtonsoft `JToken`s or `System.Text.Json`-backed JSDM values.

### CODE REVIEW

File: src/omnijson/OmniJson/OmniJsonObject.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            _wrapped = new Dictionary<string,object>();
+        }
+
+        public bool IsEmpty() => _wrapped.Count == 0;
+
+        public override OmniJsonToken this[object index] { get => this[(string)index]; set => this[(string)index] = value; }
+
+        public virtual OmniJsonToken this[string key] { get => _wrapped.ContainsKey(key) ? Wrap(_wrapped[key]) : null; set => _wrapped[key] = value?.Unwrap(); }
+
+        public virtual ICollection<string> Keys => throw new NotImplementedException();
```
Comment (asbjornu) [inline src/omnijson/OmniJson/OmniJsonObject.cs:30]: Better to not implement `IDictionary` than to throw `NotImplementedException`, no?

**Review by asbjornu — COMMENTED** [review]
> * Lost support for .NET Standard 1.x (even 1.3) because of reflection requirements in `TinyJson`.

Not a huge loss, imho. 

> This can be overcome.

How?

>  * Many tests currently failing. I had almost all of the conformance cases passing with `OmniJson` in a previous commit. I'll go back and make a pass to fix up broken cases once I get the Newtonsoft implementation finished up.

👍🏼 

> This change removes the `Newtonsoft.Json` dependency from the main project, and replaces it with a home-built facility called 'OmniJson' that provides a similar interface which supports dependency-free JSON by way of `TinyJson`, or can be enhanced with additional NuGet mix-ins (possibly called `json-ld.net.Newtonsoft` and `json-ld.net.SystemTextJson`) that would override `OmniJsonToken` et al, and provide API methods that expect and return Newtonsoft `JToken`s or `System.Text.Json`-backed JSDM values.

Besides the naming and some nits regarding the exposed interface, I think this is great. I like that we take ownership of our exposed API and are able to reduce its surface considerably.

Regarding the naming, I think the fact that the types reside within then `JsonLD` namespace is enough to differentiate them from other serializers and can thus just give the classes names such as `JsonArray`, `JsonObject`, etc.

File: src/omnijson/OmniJson/OmniJsonArray.cs [hunk 2]
```diff
@@ -0,0 +1,155 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace JsonLD.OmniJson
+{
+    public class OmniJsonArray : OmniJsonToken, IList<OmniJsonToken>, IList
```
[the remark this answers is not in the document]
Comment (sblom) [reply src/omnijson/OmniJson/OmniJsonArray.cs:8]: Really good point about just calling it `JsonLD.Json{Token|Object|Array|Value}`. That also helps get me over the other hang-up I had around how much of the Newtonsoft `JToken` interface(s) I needed to clone. (Exactly to your points below.)

**Comment by sblom** [comment]
I haven't touched this in a while, but I still believe in the approach. Let me find some time in the next several weeks to update my draft with @asbjornu's feedback.

**Pull request closed without merging**